### PR TITLE
feat: Line and block comment handling support

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: clients/vscode-hlasmplugin/src/commentEditorCommands.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+function isolateBlocks(block_candidates: CodeBlock[]): CodeBlock[] {
+    let last_last = -1;
+    const blocks: CodeBlock[] = [];
+
+    for (const b of block_candidates) {
+        if (b.first > last_last)
+            blocks.push(b);
+        else
+            blocks[blocks.length - 1].last = Math.max(b.last, last_last);
```
Comment (jirimosinger) [inline clients/vscode-hlasmplugin/src/commentEditorCommands.ts:123]: Array.prototype.at() can also be used if you find it better:
```suggestion
            blocks.at(-1).last = Math.max(b.last, last_last);
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at

File: clients/vscode-hlasmplugin/src/commentEditorCommands.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    const lineWithStatus = lines.map(x => { return { lineno: x, commented: isCommented(editor.document, x) }; })
+
+    for (const { lineno, commented } of lineWithStatus) {
+        switch (args) {
+            case CommentOption.toggle:
+                if (commented)
+                    removeComment(editor, edit, lineno);
+                else
+                    addComment(editor, edit, lineno);
+                break;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/commentEditorCommands.ts:95]: One reason is labels. Each additional '*' would remove a character from labels making the action more and more non-reversible. And we cannot reliably just move the whole line due to fixed-length file layout.

File: clients/vscode-hlasmplugin/src/test/suite/commentEditorCommands.test.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        lineCommentCommand(editor, edit, CommentOption.toggle);
+        document.text = edit.text;
+        assert.strictEqual(document.text, '   this');
+
+        lineCommentCommand(editor, edit, CommentOption.toggle);
+        document.text = edit.text;
+        assert.strictEqual(document.text, '*  this');
+    })
+
+    test('Toggle block comment', () => {
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply clients/vscode-hlasmplugin/src/test/suite/commentEditorCommands.test.ts:128]: I think this is somewhat connected to the problem of determining which lines are selected. Should the line where the selection ends be included in the selection area?

[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/test/suite/commentEditorCommands.test.ts:128]: I think so, yeah.

However, it seems to me that the problem might lie in the lines that are selected after the second toggle (used for block uncommenting). I guess an invisible character on a new line is also selected which triggers the weird behavior. Here is an additional scenario:

1. Have a new file with the following content:
```
   this
   that
   those
   these

```
2. Press CTRL+A
3. Toggle block comment e.g. 11 times and get this result:
```
         AGO   .SKIP_0_8
   this
   that
   those
   these





.SKIP_0_8 ANOP
```

New line is added for every block comment

File: clients/vscode-hlasmplugin/src/commentEditorCommands.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/commentEditorCommands.ts:95]: Sure, I get that we cannot simulate the exact same behavior as for C++. However, do you think the following is feasible/useful?

```
   this
*  that
   these
```
->
```
*  this
*  that
*  these
```
->
```
   this
   that
   these
```

[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/commentEditorCommands.ts:95]: Well, I'd probably be more happy to see this behavior for comment toggling as it feels less "random". But then again, I'm not a HLASM developer and maybe the current behavior is useful and expected by them 🙂

**Review by jirimosinger — APPROVED** [review]

**Pull request merged**